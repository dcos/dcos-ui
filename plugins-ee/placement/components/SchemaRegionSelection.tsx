import * as React from "react";

import RegionSelection from "./RegionSelection";

class SchemaRegionSelection extends React.Component {
  render() {
    const { name, formData, onChange, onBlur, onFocus } = this.props.fieldProps;

    const selectProps = {
      name,
      type: "text",
      value: formData,
      onChange: event => onChange(event.target.value),
      onBlur: onBlur && (event => onBlur(name, event.target.value)),
      onFocus: onFocus && (event => onFocus(name, event.target.value))
    };

    return <RegionSelection selectProps={selectProps} />;
  }
}

export default SchemaRegionSelection;
