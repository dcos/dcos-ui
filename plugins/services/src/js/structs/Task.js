import Item from '../../../../../src/js/structs/Item';
import ServiceImages from '../constants/ServiceImages';

module.exports = class Task extends Item {
  getId() {
    return this.get('id');
  }

  getName() {
    return this.get('name');
  }

  getImages() {
    return ServiceImages.NA_IMAGES;
  }

};
