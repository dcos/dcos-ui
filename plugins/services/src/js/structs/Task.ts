import Item from "#SRC/js/structs/Item";
import ServiceImages from "../constants/ServiceImages";

export default class Task extends Item {
  getId() {
    return this.get("id") || "";
  }

  getImages() {
    return ServiceImages.NA_IMAGES;
  }

  getName() {
    return this.get("name");
  }
}
