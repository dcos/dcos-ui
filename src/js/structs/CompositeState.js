import NodesList from './NodesList';
import ServicesList from '../../../plugins/services/src/js/structs/ServicesList';

const BLANK_NODE = {
  health: 3
};

let mergeData = function (newData, data) {
  Object.keys(newData).forEach(function (key) {
    if (Array.isArray(newData[key])) {
      data[key] = mergeMesosArrays(newData, data, key);
    } else if ((typeof (newData[key]) === 'object') && data[key]) {
      // We need to recurse over any nested objects.
      data[key] = mergeData(newData[key], data[key]);
    } else {
      // Any other type of value can be replaced.
      data[key] = newData[key];
    }
  });

  return data;
};

let mergeMesosArrays = function (newData, data, key) {
  if (key === 'frameworks' || key === 'slaves') {
    // We need to merge the objects within the frameworks and slaves arrays.
    return mergeObjectsById(newData[key], data[key]);
  } else {
    // We can replace any other array.
    return newData[key];
  }
};

let mergeObjectsById = function (newData, data = []) {
  // Merge the incoming data with the old data.
  return newData.map(function (newDatum) {
    let oldDatum = data.find(function (datum) {
      return datum.id === newDatum.id;
    });

    // These objects don't need to be deeply merged.
    return Object.assign({}, oldDatum, newDatum);
  });
};

class CompositeState {
  constructor(data = {}) {
    this.data = data;
  }

  addMarathonApps(data) {
    if (Object.keys(this.data).length === 0) {
      return;
    }

    this.data.frameworks.forEach(function (service) {
      // Marathon data merged by service name because Marathon doesn't know id.
      // See MarathonStore.processMarathonGroups
      if (data[service.name]) {
        service._meta = Object.assign({}, service._meta, {
          marathon: data[service.name]
        });
      }
    });
  }

  addNodeHealth(data) {
    if (data == null) {
      return;
    }

    let oldData = this.data.slaves || [];
    let dataByIP = {};

    data.forEach(function (datum) {
      dataByIP[datum.host_ip] = datum;
    });

    let newData = oldData.map(function (oldDatum) {
      let matchedNode = dataByIP[oldDatum.hostname] || BLANK_NODE;

      return Object.assign({}, oldDatum, matchedNode);
    });

    this.data.slaves = newData;
  }

  addState(data) {
    this.data = mergeData(data, this.data);
  }

  addSummary(data) {
    this.data = mergeData(data, this.data);
  }

  getServiceList() {
    return new ServicesList({
      items: this.data.frameworks
    });
  }

  getNodesList() {
    return new NodesList({items: this.data.slaves});
  }
}

module.exports = new CompositeState({frameworks: [], slaves: []});
