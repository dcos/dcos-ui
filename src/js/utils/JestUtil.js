import PropTypes from "prop-types";
import React from "react";
import { routerShape } from "react-router";
import { I18nProvider } from "@lingui/react";

import en from "#LOCALE/en/messages.js";

// Private router stub
const RouterStub = {
  push() {},
  replace() {},
  go() {},
  goBack() {},
  goForward() {},
  setRouteLeaveHook() {},
  createPath() {},
  createHref() {},
  isActive() {}
};

// Default prototype functions when mocking timezone
const defaultGetTimezoneOffset = Date.prototype.getTimezoneOffset;
const defaultToLocaleString = Date.prototype.toLocaleString;
const defaultDateTimeFormat = Intl.DateTimeFormat;

const JestUtil = {
  /**
   * Generates a callback function to a filter() call that will
   * keep only DOMElements matching the given tag name(s)
   *
   * @example <caption>How to use filterByTagName</caption>
   * const TestUtils = require('react-addons-test-utils');
   *
   * var componentInstance = TestUtils.renderIntoDocument(
   *   <MyTableComponent />
   * );
   *
   * var nameTDs = TestUtils.scryRenderedDOMComponentsWithClass(
   *   componentInstance, 'table-column-name'
   * ).filter(filterByTagName('TD'));
   *
   * @param {array|string} tag - One or mor tagNames to match
   * @returns {function} Returns a callback function to be passed on .filter()
   */
  filterByTagName(tag) {
    if (!Array.isArray(tag)) {
      tag = [tag];
    }

    return function(element) {
      return tag.some(function(tagName) {
        return element.tagName === tagName;
      });
    };
  },

  /**
   * Helper callback function to be used when you want the text contents
   * of an array of DOMElements.
   *
   * @example <caption>How to use mapTextContent</caption>
   * const TestUtils = require('react-addons-test-utils');
   *
   * var componentInstance = TestUtils.renderIntoDocument(
   *   <MyTableComponent />
   * );
   *
   * var nameTDContents = TestUtils.scryRenderedDOMComponentsWithClass(
   *   componentInstance,
   *   'table-column-name'
   * ).map(mapTextContent);
   *
   * @param {DOMElement} element - The DOM element for which to get the content
   * @returns {string} The text content
   */
  mapTextContent(element) {
    return element.textContent.trim();
  },

  /**
   * Mock a different timezone by overriding relevant Date primitive
   * prototype functions.
   *
   * This function will mock Date.getTimezoneOffset, Date.toLocaleString and Intl.DateTimeFormat
   *
   * @param {String} timezone - The IANA timezone string (ex. Europe/Athens) or 'UTC'
   */
  mockTimezone(timezone) {
    const date = new Date();
    const timezoneOffset =
      (new Date(date.toLocaleString(undefined, { timeZone: "UTC" })) -
        new Date(date.toLocaleString(undefined, { timeZone: timezone }))) /
      1000 /
      60;

    /* eslint-disable no-extend-native */
    Date.prototype.getTimezoneOffset = function() {
      return timezoneOffset;
    };
    Date.prototype.toLocaleString = function(locale = undefined, options = {}) {
      options.timeZone = options.timeZone || timezone;

      return defaultToLocaleString.call(this, locale, options);
    };
    Intl.DateTimeFormat = function(locales, options) {
      options.timeZone = options.timeZone || timezone;

      return defaultDateTimeFormat.call(this, locales, options);
    };
    /* eslint-enable no-extend-native */
  },

  /**
   * Helper to stub router context, based on
   * https://github.com/reactjs/react-router/blob/0.13.x/docs/guides/testing.md
   * @param {React.Component} Component
   * @param {object} [routerStubs]
   * @param {object} [contextTypes]
   * @param {object} [context]
   * @returns {React.Element} wrapped component element
   */
  stubRouterContext(
    Component,
    routerStubs = {},
    contextTypes = {},
    context = {}
  ) {
    // Create wrapper component
    return class WrappedComponent extends React.Component {
      static get childContextTypes() {
        return Object.assign(
          {
            router: routerShape,
            routeDepth: PropTypes.number
          },
          contextTypes
        );
      }

      getChildContext() {
        return Object.assign(
          {
            router: Object.assign(RouterStub, routerStubs),
            routeDepth: 0
          },
          context
        );
      }

      render() {
        return <Component {...this.props} />;
      }
    };
  },

  withI18nProvider(Component, catalogs = { en }) {
    // eslint-disable-next-line react/no-multi-comp
    return class WrappedComponent extends React.Component {
      render() {
        return (
          <I18nProvider defaultRender="span" language="en" catalogs={catalogs}>
            <Component {...this.props} />
          </I18nProvider>
        );
      }
    };
  },

  /**
   * Generates a callback function to a reduce() call that will
   * find and return the textContent of DOM element(s) matching
   * the given selector
   *
   * @example <caption>How to use reduceTextContentOfSelector</caption>
   * const TestUtils = require('react-addons-test-utils');
   *
   * var componentInstance = TestUtils.renderIntoDocument(
   *   <MyTableComponent />
   * );
   *
   * var names = TestUtils.scryRenderedDOMComponentsWithClass(
   *   componentInstance,
   *   'table-column-name'
   * ).reduce(
   *   reduceTextContentOfSelector('.collapsing-string-full-string'),
   *   []
   * );
   *
   * @param {string} selector - The CSS selector to match children
   * @returns {function} Returns a callback function to be passed on .reduce()
   */
  reduceTextContentOfSelector(selector) {
    return function(strings, element) {
      const matchedElements = Array.from(element.querySelectorAll(selector));

      matchedElements.forEach(function(stringElement) {
        strings.push(stringElement.textContent);
      });

      return strings;
    };
  },

  /**
   * Restore the original version of the Date prototype functions, overwritten
   * by the mockTimezone function.
   */
  unmockTimezone() {
    /* eslint-disable no-extend-native */
    Date.prototype.getTimezoneOffset = defaultGetTimezoneOffset;
    Date.prototype.toLocaleString = defaultToLocaleString;
    Intl.DateTimeFormat = defaultDateTimeFormat;
    /* eslint-enable no-extend-native */
  }
};

module.exports = JestUtil;
