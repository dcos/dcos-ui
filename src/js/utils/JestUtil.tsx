import PropTypes from "prop-types";
import * as React from "react";
import { routerShape } from "react-router";
import { I18nProvider } from "@lingui/react";

import en from "#LOCALE/en/messages.js";

// Private router stub
const RouterStub = {
  push() {},
  replace() {},
  go() {},
  goBack() {},
  goForward() {},
  setRouteLeaveHook() {},
  createPath() {},
  createHref() {},
  isActive() {},
};

// Default prototype functions when mocking timezone
const defaultGetTimezoneOffset = Date.prototype.getTimezoneOffset;
const defaultToLocaleString = Date.prototype.toLocaleString;
const defaultDateTimeFormat = Intl.DateTimeFormat;

const JestUtil = {
  /**
   * Mock a different timezone by overriding relevant Date primitive
   * prototype functions.
   *
   * This function will mock Date.getTimezoneOffset, Date.toLocaleString and Intl.DateTimeFormat
   *
   * @param {String} timezone - The IANA timezone string (ex. Europe/Athens) or 'UTC'
   */
  mockTimezone(timezone) {
    const date = new Date();
    const timezoneOffset =
      (new Date(date.toLocaleString(undefined, { timeZone: "UTC" })) -
        new Date(date.toLocaleString(undefined, { timeZone: timezone }))) /
      1000 /
      60;

    Date.prototype.getTimezoneOffset = () => timezoneOffset;
    Date.prototype.toLocaleString = function (
      locale = undefined,
      options = {}
    ) {
      options.timeZone = options.timeZone || timezone;

      return defaultToLocaleString.call(this, locale, options);
    };
    Intl.DateTimeFormat = function (locales, options) {
      options.timeZone = options.timeZone || timezone;

      return defaultDateTimeFormat.call(this, locales, options);
    };
  },

  /**
   * Helper to stub router context, based on
   * https://github.com/reactjs/react-router/blob/0.13.x/docs/guides/testing.md
   * @param {React.Component} Component
   * @param {object} [routerStubs]
   * @param {object} [contextTypes]
   * @param {object} [context]
   * @returns {React.Element} wrapped component element
   */
  stubRouterContext(
    Component,
    routerStubs = {},
    contextTypes = {},
    context = {}
  ) {
    // Create wrapper component
    return class WrappedComponent extends React.Component {
      static get childContextTypes() {
        return {
          router: routerShape,
          routeDepth: PropTypes.number,
          ...contextTypes,
        };
      }

      getChildContext() {
        return {
          router: { ...RouterStub, ...routerStubs },
          routeDepth: 0,
          ...context,
        };
      }

      render() {
        return <Component {...this.props} />;
      }
    };
  },

  withI18nProvider(Component, catalogs = { en }) {
    return class WrappedComponent extends React.Component {
      render() {
        return (
          <I18nProvider defaultRender="span" language="en" catalogs={catalogs}>
            <Component {...this.props} />
          </I18nProvider>
        );
      }
    };
  },

  /**
   * Restore the original version of the Date prototype functions, overwritten
   * by the mockTimezone function.
   */
  unmockTimezone() {
    Date.prototype.getTimezoneOffset = defaultGetTimezoneOffset;
    Date.prototype.toLocaleString = defaultToLocaleString;
    Intl.DateTimeFormat = defaultDateTimeFormat;
  },
};

export default JestUtil;
