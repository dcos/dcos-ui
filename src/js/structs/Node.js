import Item from "./Item";
import TaskStates from "../../../plugins/services/src/js/constants/TaskStates";
import UnitHealthUtil from "../utils/UnitHealthUtil";
import { findNestedPropertyInObject } from "../utils/Util";

class Node extends Item {
  getID() {
    return this.get("id");
  }

  getServiceIDs() {
    return this.get("framework_ids");
  }

  isActive() {
    return this.get("active");
  }

  getDomain() {
    return this.get("domain");
  }

  getRegionName() {
    const nodeRegionName = findNestedPropertyInObject(
      this.getDomain(),
      "fault_domain.region.name"
    );

    return nodeRegionName || "N/A";
  }

  getZoneName() {
    const nodeZoneName = findNestedPropertyInObject(
      this.getDomain(),
      "fault_domain.zone.name"
    );

    return nodeZoneName || "N/A";
  }

  getUsageStats(resource) {
    const total = (this.get("resources") || {})[resource];
    const value = (this.get("used_resources") || {})[resource];

    if (total === undefined || value === undefined) {
      return { percentage: 0, total: 0, value: 0 };
    }

    const percentage = Math.round((100 * value) / Math.max(1, total));

    return { percentage, total, value };
  }

  getHostName() {
    return this.get("hostname");
  }

  // Below is Component Health specific API
  // http://schema.dcos/system/health/node
  getHealth() {
    return UnitHealthUtil.getHealth(this.get("health"));
  }

  getOutput() {
    if (typeof this.get("output") === undefined) {
      return "N/A";
    }

    return this.get("output") || "OK";
  }

  sumTaskTypesByState(state) {
    let sum = 0;

    Object.keys(TaskStates).forEach(function(taskType) {
      if (
        TaskStates[taskType] &&
        TaskStates[taskType].stateTypes.indexOf(state) !== -1
      ) {
        // Make sure there's a value
        if (this[taskType]) {
          sum += this[taskType];
        }
      }
    }, this);

    return sum;
  }

  getResources() {
    return (
      this.get("used_resources") || {
        cpus: 0,
        mem: 0,
        gpus: 0,
        disk: 0
      }
    );
  }

  isPublic() {
    return (
      findNestedPropertyInObject(this.get("attributes"), "public_ip") === "true"
    );
  }

  getIp() {
    return this.get("host_ip") || this.getHostName();
  }

  getPublicIps() {
    if (!this.get("network")) {
      return [];
    }

    return this.get("network").public_ips || [];
  }
}

module.exports = Node;
