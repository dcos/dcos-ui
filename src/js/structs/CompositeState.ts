import NodesList from "./NodesList";
import Node from "./Node";
import Util from "../utils/Util";
import ServicesList from "../../../plugins/services/src/js/structs/ServicesList";

const MISSING_HEALTH_NODE = {
  health: 3,
};

// Mutates node data with node health data. This is much faster than mapping and creating new node objects
const enrichNodeDataWithHealthData = (nodes, healthData) => {
  nodes.forEach((node) => {
    const matchedHealthNode = healthData[node.hostname] || MISSING_HEALTH_NODE;

    node.health = matchedHealthNode.health;
  });
};

let storedNodeHealth, storedState;
class CompositeState {
  constructor(data = {}) {
    this._refCount = 0;
    this.data = data;
    this.masterInfo = null;
    this.nodeHealthData = {};
  }

  /**
   * Enables the composite state if there is no
   * one requesting it to be disabled anymore
   */
  enable() {
    this._refCount = Math.max(this._refCount - 1, 0);

    // if this update enables the composite state again, let's update everyting
    if (!this._isDisabled()) {
      if (storedNodeHealth) {
        this.addNodeHealth(storedNodeHealth);
        storedNodeHealth = null;
      }

      if (storedState) {
        this.addState(storedState);
        storedState = null;
      }
    }
  }

  /**
   * Disables the composite state
   */
  disable() {
    this._refCount = this._refCount + 1;
  }

  _isDisabled() {
    return this._refCount > 0;
  }

  addNodeHealth(data) {
    if (this._isDisabled()) {
      storedNodeHealth = data;

      return;
    }
    if (data == null) {
      return;
    }

    // Memoize node health data as an object with "host_ip" keys
    this.nodeHealthData = Util.keyBy(data, "host_ip");

    this.data.slaves = this.data.slaves || [];
    enrichNodeDataWithHealthData(this.data.slaves, this.nodeHealthData);
  }

  addMasterInfo(masterInfo) {
    this.masterInfo = masterInfo;
  }

  addState(newData) {
    if (this._isDisabled()) {
      storedState = newData;

      return;
    }
    if (newData == null) {
      return;
    }

    if (this.masterInfo) {
      newData.master_info = this.masterInfo;
    }
    this.data = newData;

    // Reuse memoized node health data
    this.data.slaves = this.data.slaves || [];
    enrichNodeDataWithHealthData(this.data.slaves, this.nodeHealthData);
  }

  getServiceList() {
    return new ServicesList({
      items: this.data.frameworks,
    });
  }

  set data(newData) {
    this._data = newData;
    this._nodesList = null;
  }

  get data() {
    return this._data;
  }

  getNodesList() {
    if (this._nodesList === null) {
      this._nodesList = new NodesList({ items: this.data.slaves });
    }

    return this._nodesList;
  }

  getMasterNode() {
    return new Node(this.data.master_info);
  }
}

export default new CompositeState({ frameworks: [], slaves: [] });
