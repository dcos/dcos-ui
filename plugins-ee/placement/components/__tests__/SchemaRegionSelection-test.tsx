import CompositeState from "#SRC/js/structs/CompositeState";
import Node from "#SRC/js/structs/Node";
import Task from "#PLUGINS/services/src/js/structs/Task";
import JestUtil from "#SRC/js/utils/JestUtil";
import SchemaRegionSelection from "../SchemaRegionSelection";

jest.mock("#SRC/js/structs/CompositeState");

const React = require("react");

const renderer = require("react-test-renderer");
const MesosStateStore = require("#SRC/js/stores/MesosStateStore").default;

const MasterNodeLocal = require("./fixtures/MasterNodeLocal.json");

let thisInstance;

describe("SchemaRegionSelection", () => {
  beforeEach(() => {
    CompositeState.getMasterNode = () => new Node(MasterNodeLocal);

    MesosStateStore.addChangeListener = () => {};
    MesosStateStore.getTaskFromTaskID = () =>
      new Task({
        id: "bar",
        state: "TASK_RUNNING"
      });
    // Create mock functions
    MesosStateStore.get = key => {
      if (key === "lastMesosState") {
        return {
          slaves: [
            {
              domain: {
                fault_domain: {
                  region: { name: "us-west-2" },
                  zone: { name: "us-west-2b" }
                }
              }
            }
          ]
        };
      }
    };
  });

  describe("#localRegionOptions", () => {
    it("attaches local to option text when region is equal to master", () => {
      const fieldProps = {
        name: "foo",
        type: "text",
        value: {},
        onChange: () => {},
        onBlur: () => {},
        onFocus: () => {}
      };
      const WrappedComponent = JestUtil.withI18nProvider(SchemaRegionSelection);
      thisInstance = renderer.create(
        <WrappedComponent fieldProps={fieldProps} />
      );

      const tree = thisInstance.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
