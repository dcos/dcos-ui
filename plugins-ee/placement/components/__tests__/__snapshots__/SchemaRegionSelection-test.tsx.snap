// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SchemaRegionSelection #localRegionOptions attaches local to option text when region is equal to master 1`] = `
<div
  className="form-row row"
>
  <div
    className="form-group"
  >
    <label
      className="column-12"
    >
      <div
        className="form-group-heading"
      >
        <div
          className="form-group-heading-content form-group-heading-content-primary"
          title="Region"
        >
          <span />
        </div>
      </div>
    </label>
    <div
      className="column-5"
    >
      <span
        className="form-control form-control-select"
      >
        <select
          name="foo"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="text"
          value=""
        >
          <option
            value=""
          >
            Select...
          </option>
          <option
            value="us-west-2"
          >
            us-west-2 (Local)
          </option>
        </select>
      </span>
    </div>
    <p
      className="form-control-feedback column-12 text-no-transform"
    >
      <span />
    </p>
  </div>
</div>
`;
